### PR TITLE
Fix computeViewRectangle accuracy

### PR DESCRIPTION
**Pull request by njrivera**
Closes https://github.com/CesiumGS/cesium/issues/9516

Background:
The issue with the view rectangle when the horizon is visible is that the horizon quad, while correct, assumes that the heading of the camera is always toward the north.
- Suppose the camera is pointed south and angled up such that the horizon is visible on the UL and UR corners of the camera
  - It's UL and UR corners (SE and SW respectively) would be replaced by the horizon quad's UL and UR corners
  - But the horizon quad assumes the camera direction is north, so its UL and UR corners are really NW and NE, which means that the view rectangle is computed much higher than should be

The fix involves 'rotating' the elements of the horizon quad to match the camera's corners so that the correct corners are replaced when 1 or 2 corners do not have successful picks (are showing the horizon)

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @njrivera!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/fix-computeViewRectangle-accuracy/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

**Comment by ebogo1** [comment]
@njrivera Thanks for the fix - the code changes look great and I will take a look in the next few days. Just pinging to let you know I'm looking at it since the Camera code can be finicky 👍

File: Source/Scene/Camera.js [hunk 1]
```diff
@@ -3681,7 +3681,60 @@ var cartoArray = [
...
+  var newQuad = new Array(4);
+  for (var i = 0, j = numRotations; i < 4; i++, j = (j + 1) % 4) {
+    newQuad[j] = quad[i];
+  }
+  return newQuad;
+}
+
+function generalHeading(heading) {
+  if (heading > 45 && heading <= 135) {
+    return "east";
```
Comment (ebogo1) [inline Source/Scene/Camera.js:3695]: Just a minor nit, but I don't think we usually do returns of strings like this. Maybe an enum for this function is overkill though..

File: Source/Scene/Camera.js [hunk 2]
```diff
@@ -3727,48 +3821,30 @@ Camera.prototype.computeViewRectangle = function (ellipsoid, result) {
...
+  }
 
-  if (successfulPickCount < 2) {
+  if (unsuccessfulPicks.length > 2) {
     // If we have space non-globe in 3 or 4 corners then return the whole globe
     return Rectangle.MAX_VALUE;
   }
 
+  if (unsuccessfulPicks.length > 0) {
+    adjustViewForHorizon(this, unsuccessfulPicks, ellipsoid);
```
Comment (ebogo1) [inline Source/Scene/Camera.js:3845]: There should be an entry in `CHANGES.md` under the "Fixes" section for this.

File: Source/Scene/Camera.js [hunk 3]
```diff
@@ -3681,7 +3681,60 @@ var cartoArray = [
...
+// and adjusts the quad array so that its corners roughly match where the camera is pointing.
+//
+// Ex.
+//    - direction is 'left' (looking toward the west with the horizon showing on the upper and lower left corners)
+//    - heading is 80 - camera is rotated clockwise toward the east (still showing horizon on upper and lower left corners).
+//                      At this point, the camera is upside down (up -> south).
+//
+// Given that the corner indeces of the quad are [UL, LL, LR, UR],
+// the quad elements would be 'rotated' like so - [NW, SW, SE, NE] => [SE, NE, NW, SW]
+function changeQuadHeading(heading, direction, quad) {
```
Comment (ebogo1) [inline Source/Scene/Camera.js:3731]: This makes sense to me but I wonder if it's a rigorous solution. I might just need some time to think about it some more, but I haven't wrapped my head around adding the number of rotations to the general heading count yet.

@lilleyse what do you think about this math?

**Review by ebogo1 — COMMENTED** [review]
Thanks @njrivera! I confirmed that the [Sandcastle](https://sandcastle.cesium.com/#c=pZJNj5swEIb/ipUTSHSAQj5ISdQ27a1VpWzVExcHZhOrxqa2IUpX+e+1cTb7qb3UFzy233fmGWagigwMj6jIigg8kg1q1rfwazwLqkk9xhspDGUCVTWJyB1RKBpUW/zTozbfZYNLYlSP5Bx+qMRgLWvaoqLW0nuDj+1lJfwWOqmZYVI8TbuhytgdFVmQFbM8KSBN5xF5N51nWbJIYJ7MIpK/z4v5tIBFNnP5LoYNU1i/6ZhAlhYRSSCdptYzgSKbhld9370lzPOLYjF+kywPHYxjRWGYOT2wjjFDDbRpgrtKELtcaVTsOZIluRy5VUupGiaoTWQvnvQKatl2vUH3I7b36uDyRNcoEPZc7hCQc9ZpyZowejBuraVilC+vNJJLBduvX+DIzOET7w7UYYVecq7EeeQhN1Q0NdWGoyv/p5R8R9Xn3hgp7DBsfE3EFUWuVbmZuO2Fb34QekDfFbiCw2PW1X+z2hTjsE2iSanNieP6Hv4jazupDOkVDwBig23HXdJ419e/0UCttRO6p2X8WFo2bCCsWb0y86TmVGt7c9tzfsP+WuR1Gdv3L6RcUgu5/zGg4vTknh3S9Td/CABlbMPXlcb3+pnzPw) from the original issue looks way better with these changes. And it seems to work in all four directions from my testing. 

The approach looks good to me, and I think it's correct to adjust the view based on the number of unsuccessful horizon picks since this issue only comes up when the horizon is visible. I need to test 2D and CV with these changes too.

I had a few questions/comments, cc @lilleyse for any other thoughts -

File: Source/Scene/Camera.js (hunk 1 above)
Comment (njrivera) [reply Source/Scene/Camera.js]: I removed the strings and just returned the `headingQuadrant` enum values

File: Source/Scene/Camera.js (hunk 2 above)
Comment (njrivera) [reply Source/Scene/Camera.js]: Added

**Comment by ggetz** [comment]
@lilleyse Would you be able to do a sanity check here on the math? Looks like this PR is in pretty good shape and it would be great to move it along.

**Comment by lilleyse** [comment]
I spent a little time looking at this a few weeks ago, and if @sanjeetsuhag is picking this up now here's where I left off - I noticed that `DepthPlane` and `computeViewRectangle` have similar functionality for computing the cartesian polygon (e.g. both have a comment `// Determine the radius of the 'limb' of the ellipsoid`). Might want to double check that the math in `computeViewRectangle` is the same, and then figure out the optimal way to convert a cartesian polygon to a cartographic rectangle.

File: Source/Scene/Camera.js [hunk 4]
```diff
@@ -3711,13 +3763,53 @@ function addToResult(x, y, index, camera, ellipsoid, computedHorizonQuad) {
...
+  return false;
+}
+
+// Indeces in a quad array that represents a rectangle
+var UL = 0;
+var LL = 1;
+var LR = 2;
+var UR = 3;
+
+function generalCameraDirection(cornersShowingHorizon) {
```
Comment (sanjeetsuhag) [inline Source/Scene/Camera.js:3777]: I've tested each case here and it seems to work as advertised.

File: Source/Scene/Camera.js [hunk 5]
```diff
@@ -3701,7 +3701,59 @@ var cartoArray = [
...
+
+var headingQuadrant = {
+  north: 0,
+  east: 1,
+  south: 2,
+  west: 3,
+};
+
+function generalHeading(heading) {
+  if (heading > 45 && heading <= 135) {
```
Comment (sanjeetsuhag) [inline Source/Scene/Camera.js:3721]: Once this is converted to radians, `CesiumMath` constants can be used here.

**Review by sanjeetsuhag — COMMENTED** [review]
I've tested this pretty thoroughly now and it's working really well in 3D view. In 2D view, I think I'm running into https://github.com/CesiumGS/cesium/issues/4346 because a lot of simple cases are broken.

**Comment by ggetz** [comment]
@njrivera We're trying to be more responsive to PRs and keep our PR backlog tidy. Due to inactivity, I'm going to close this PR. But please feel free to re-open when you get the chance to work on this! Thanks for your patience.

**Pull request closed without merging**